### ISSUE TITLE
arguments are located on different GPUs

### ISSUE THREAD

**Issue by a791702141** [pytorch/pytorch#4067]
train.lua

if opt.gpu > 0 then
   require 'cunn'
   cutorch.setDevice(opt.gpu)
   --cutorch.setDevice(opt.gpu1)
   input_ctx_vis = input_ctx_vis:cuda(); input_ctx = input_ctx:cuda();  input_center = input_center:cuda()
   noise = noise:cuda();  label = label:cuda()
   netG = util.cudnn(netG);     netD = util.cudnn(netD)
   netD:cuda();           netG:cuda();           criterion:cuda();      
   if opt.wtl2~=0 then
      criterionMSE:cuda(); input_real_center = input_real_center:cuda();
   end
end
print('NetG:',netG)
print('NetD:',netD)

local parametersD, gradParametersD = netD:getParameters()
local parametersG, gradParametersG = netG:getParameters()

if opt.display then disp = require 'display' end

noise_vis = noise:clone()
if opt.noisetype == 'uniform' then
    noise_vis:uniform(-1, 1)
elseif opt.noisetype == 'normal' then
    noise_vis:normal(0, 1)
end

---------------------------------------------------------------------------
-- Define generator and adversary closures
---------------------------------------------------------------------------
-- create closure to evaluate f(X) and df/dX of discriminator
local fDx = function(x)
   netD:apply(function(m) if torch.type(m):find('Convolution') then m.bias:zero() end end)
   netG:apply(function(m) if torch.type(m):find('Convolution') then m.bias:zero() end end)

   gradParametersD:zero()

   -- train with real
   data_tm:reset(); data_tm:resume()
   local real_ctx = data:getBatch()
   local real_center = real_ctx[{{},{},{1 + opt.fineSize/4, opt.fineSize/2 + opt.fineSize/4},{1 + opt.fineSize/4, opt.fineSize/2 + opt.fineSize/4}}]:clone() -- copy by value
   real_ctx[{{},{1},{1 + opt.fineSize/4 + opt.overlapPred, opt.fineSize/2 + opt.fineSize/4 - opt.overlapPred},{1 + opt.fineSize/4 + opt.overlapPred, opt.fineSize/2 + opt.fineSize/4 - opt.overlapPred}}] = 2*117.0/255.0 - 1.0
   real_ctx[{{},{2},{1 + opt.fineSize/4 + opt.overlapPred, opt.fineSize/2 + opt.fineSize/4 - opt.overlapPred},{1 + opt.fineSize/4 + opt.overlapPred, opt.fineSize/2 + opt.fineSize/4 - opt.overlapPred}}] = 2*104.0/255.0 - 1.0
   real_ctx[{{},{3},{1 + opt.fineSize/4 + opt.overlapPred, opt.fineSize/2 + opt.fineSize/4 - opt.overlapPred},{1 + opt.fineSize/4 + opt.overlapPred, opt.fineSize/2 + opt.fineSize/4 - opt.overlapPred}}] = 2*123.0/255.0 - 1.0
   input_ctx:copy(real_ctx)
   input_center:copy(real_center)
   data_tm:stop()
   if opt.wtl2~=0 then
      input_real_center:copy(real_center)
   end
   label:fill(real_label)

   local output
   if opt.conditionAdv then
      output = netD:forward({input_ctx,input_center})
   else
      output = netD:forward(input_center)
   end
   local errD_real = criterion:forward(output, label)
   local df_do = criterion:backward(output, label)
   if opt.conditionAdv then
      netD:backward({input_ctx,input_center}, df_do)
   else
      netD:backward(input_center, df_do)
   end

and util.lua

function util.cudnn(net)
    -- assume cunn and cudnn are both avaliable
    -- use cudnn convolution layer
    --recursiveCudnn(net)
    cudnn.convert(net, cudnn)

    --local nGPU = cutorch.getDeviceCount()
    local nGPU = 2
    -- support data parallel for multi gpus

    if nGPU > 1 then
        local parallel_net = nn.DataParallelTable(1)
        parallel_net:add(net,{5,6})--:cuda()
        return parallel_net
      --  local gpus = torch.range(5, nGPU):totable()
       -- print(gpus)
        --local fastest, benchmark = cudnn.fastest, cudnn.benchmark
        --local parallel_net = nn.DataParallelTable(1, true, true)
        --parallel_net:add(net, gpus):cuda
                --:threads(function()
                   -- local cudnn = require 'cudnn' 
                   -- cudnn.fastest, cudnn.benchmark = fastest, benchmark
                --end)
        --return parallel_net
    else
        return net
    end
end

When I executed I got this 

/usr/local/torch/install/bin/luajit: train.lua:276: arguments are located on different GPUs at /usr/local/torch/extra/cutorch/lib/THC/generic/THCTensorMath.cu:21
stack traceback:
	[C]: in function 'zero'
	train.lua:276: in function 'callback'
	/usr/local/torch/install/share/lua/5.1/nn/Module.lua:352: in function 'apply'
	/usr/local/torch/install/share/lua/5.1/nn/Module.lua:356: in function 'apply'
	/usr/local/torch/install/share/lua/5.1/nn/Module.lua:356: in function 'apply'
	...l/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:464: in function 'apply'
	train.lua:276: in function 'opfunc'
	/usr/local/torch/install/share/lua/5.1/optim/adam.lua:37: in function 'adam'
	train.lua:413: in main chunk
	[C]: in function 'dofile'
	...ocal/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
	[C]: at 0x004064f0

**Comment by a791702141** [comment]
There are two Titan X GPU. And I can got the right answer using each of them. But when I tried to combine them, I failed and g ot    this message

**Comment by zou3519** [comment]
You might want to open an issue in [cutorch](https://github.com/torch/cutorch) instead of here.

**Comment by gurkirt** [comment]
You might want to visit https://discuss.pytorch.org/t/index-select-with-dataparallel-arguments-are-located-on-different-gpus/10229/7.

I am not sure if you provided the whole full script. Just be careful when using module list. If you net contains a list of modules in a list then you might want to associate your modules to base net self-object. Or use nn.Modulelist(list_of_modules).

**Comment by ChenSongle** [comment]
I also met this issue. I solved this problem by put the below code on the beginning of the program.

if opt.cuda then
      print"Using CUDA"
      require 'cutorch'
      require 'cunn'
      cutorch.setDevice(opt.useDevice)
end
